cmd/snap/cmd_info.go

```diff
@@ -0,0 +1,191 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package main
+
+import (
+	"fmt"
+	"io"
+	"path/filepath"
+	"text/tabwriter"
+
+	"github.com/jessevdk/go-flags"
+
+	"github.com/snapcore/snapd/client"
+	"github.com/snapcore/snapd/i18n"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/snap"
+)
+
+type infoCmd struct {
+	Verbose    bool `long:"verbose"`
+	Positional struct {
+		Snaps []string `positional-arg-name:"<snap>" required:"1"`
+	} `positional-args:"yes" required:"yes"`
+}
+
+var shortInfoHelp = i18n.G("show detailed information about a snap")
+var longInfoHelp = i18n.G(`
+The info command shows detailed information about a snap, be it by name or by path.`)
+
+func init() {
+	addCommand("info",
+		shortInfoHelp,
+		longInfoHelp,
+		func() flags.Commander {
+			return &infoCmd{}
+		}, map[string]string{
+			"verbose": i18n.G("Include a verbose list of a snap's notes (otherwise, summarise notes)"),
+		}, nil)
+}
+
+func norm(path string) string {
+	path = filepath.Clean(path)
+	if osutil.IsDirectory(path) {
+		path = path + "/"
+	}
+
+	return path
+}
+
+func tryDirect(w io.Writer, path string, verbose bool) bool {
+	path = norm(path)
+
+	snapf, err := snap.Open(path)
+	if err != nil {
+		return false
+	}
+
+	info, err := snap.ReadInfoFromSnapFile(snapf, nil)
+	if err != nil {
+		return false
+	}
+	fmt.Fprintf(w, "path:\t%q\n", path)
+	fmt.Fprintf(w, "name:\t%s\n", info.Name())
+	fmt.Fprintf(w, "summary:\t%q\n", info.Summary())
+
+	var notes *Notes
+	if verbose {
+		fmt.Fprintln(w, "notes:\t")
+		fmt.Fprintf(w, "  confinement:\t%s\n", info.Confinement)
+		if info.Broken == "" {
+			fmt.Fprintln(w, "  broken:\tfalse")
+		} else {
+			fmt.Fprintf(w, "  broken:\ttrue (%s)\n", info.Broken)
+		}
+
+	} else {
+		notes = NotesFromInfo(info)
+	}
+	fmt.Fprintf(w, "version:\t%s %s\n", info.Version, notes)
+	if info.Type != snap.TypeApp {
+		fmt.Fprintf(w, "type:\t%s\n", info.Type)
+	}
+
+	return true
+}
+
+func coalesce(snaps ...*client.Snap) *client.Snap {
+	for _, s := range snaps {
+		if s != nil {
+			return s
+		}
+	}
+	return nil
+}
+
+func (x *infoCmd) Execute([]string) error {
+	cli := Client()
+
+	w := tabwriter.NewWriter(Stdout, 2, 2, 1, ' ', 0)
+
+	noneOK := true
+	for i, snapName := range x.Positional.Snaps {
+		if i > 0 {
+			fmt.Fprintln(w, "---")
+		}
+
+		if tryDirect(w, snapName, x.Verbose) {
+			noneOK = false
+			continue
+		}
+
+		remote, _, _ := cli.FindOne(snapName)
+		local, _, _ := cli.Snap(snapName)
+
+		both := coalesce(local, remote)
+
+		if both == nil {
+			fmt.Fprintf(w, "path:\t%q\nwarning:\t%s\n", snapName, i18n.G("not a valid snap"))
```

Comment:
Similarly, this is saying that something is wrong without saying what. We got the actual errors above. We should inspect and report when unexpected.

Also, do we know this is a path-related error rather than some other issue?